File: Streamfox.Server/Types/Optional.cs
namespace Streamfox.Server.Types
{
    public readonly struct Optional<T>
    {
        private Optional(T value)
        {
            HasValue = true;
            Value = value;
        }

        public bool HasValue { get; }

        public T Value { get; }

        public static Optional<T> Of(T value)
        {
            return new Optional<T>(value);
        }

        public static Optional<T> Empty()
        {
            return new Optional<T>();
        }
    }

    public static class Optional
    {
        public static Optional<T> Of<T>(T value)
        {
            return Optional<T>.Of(value);
        }
    }
}

File: frontend/src/utils/optional.ts
export type Optional<T> = T | null;

export function empty<T>(): Optional<T> {
  return null;
}

export function hasValue<T>(opt: Optional<T>): boolean {
  return opt !== null;
}

export function getValue<T>(opt: Optional<T>): T {
  return opt as T;
}
